syntax = "proto3";

package dbpb;

import "dbpb/ast.proto";

enum QueryType {
  QUERY_TYPE_UNSPECIFIED = 0;
  QUERY_TYPE_INSERT = 1;
  QUERY_TYPE_SELECT = 2;
  QUERY_TYPE_UPDATE = 3;
  QUERY_TYPE_DELETE = 4;
  QUERY_TYPE_CREATE = 5;
  QUERY_TYPE_DROP = 6;
}

message DatabaseRequest {
  string db_name = 1;
  repeated AstSubclass queries = 2;
}

message ErrorStatus {
  string message = 1;
}

message DatabaseHeaderColumn {
  string column_name = 10;
  string column_type = 11;
}

message DatabaseHeaderRow {
  QueryType query_type = 21;
  repeated DatabaseHeaderColumn columns = 22;
}

message DatabaseBodyColumn {
  string column_value = 31;
}

message DatabaseBodyRow {
  repeated DatabaseBodyColumn columns = 41;
}

message DatabaseResponse {
  ErrorStatus err_status = 50;
  oneof row {
    DatabaseHeaderRow header = 51;
    DatabaseBodyRow body = 52;
  }
}

service DatabaseQuery {
  rpc PerformQuery(DatabaseRequest) returns (DatabaseResponse) {}
  rpc PerformQuerySS(DatabaseRequest) returns (stream DatabaseResponse) {}
}
